tsl/src/continuous_aggs/refresh.c

```diff
@@ -374,3 +374,54 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
 			 "continuous aggregate \"%s\" is already up-to-date",
 			 NameStr(cagg->data.user_view_name));
 }
+
+/*
+ * Refresh all continuous aggregates on a hypertable.
+ *
+ * The refreshing happens in a single transaction. For this to work correctly,
+ * there must be no new invalidations written in the refreshed region during
+ * the refresh. Therefore, the caller is responsible for proper locking to
+ * ensure there are no invalidations.
+ */
+void
+continuous_agg_refresh_all(const Hypertable *ht, int64 start, int64 end)
+{
+	Catalog *catalog = ts_catalog_get();
+	List *caggs = ts_continuous_aggs_find_by_raw_table_id(ht->fd.id);
+	Dimension *dim = hyperspace_get_open_dimension(ht->space, 0);
+	InternalTimeRange refresh_window = {
+		.type = ts_dimension_get_partition_type(dim),
+		.start = start,
+		.end = end,
+	};
+	ListCell *lc;
+
+	LockRelationOid(catalog_get_table_id(catalog, CONTINUOUS_AGGS_INVALIDATION_THRESHOLD),
+					AccessExclusiveLock);
+	continuous_agg_invalidation_threshold_set(ht->fd.id, refresh_window.end);
+
+	foreach (lc, caggs)
+	{
+		const ContinuousAgg *cagg = lfirst(lc);
+		const Hypertable *cagg_ht = ts_hypertable_get_by_id(cagg->data.mat_hypertable_id);
+		InvalidationStore *invalidations;
+
+		if (cagg == linitial(caggs))
+		{
+			/* It is enough to process the hypertable invalidation log once,
+			 * so do it only for the first continuous aggregate */
+			invalidation_process_hypertable_log(cagg, &refresh_window);
+			/* Must make invalidation processing visible */
+			CommandCounterIncrement();
+		}
```

Comment:
Nit: It is probably clearer to move this outside the foreach loop and use `linitial` to get the first element. Then you won't need the comment either.